Testnet/MintableTokenInvoice/MintableTokenInvoice/MintableTokenInvoice.cs

```diff
@@ -69,10 +85,32 @@ private struct TransactionReferenceTemplate
         public Address address;
     }
 
-    private string GetTransactionReference(UInt128 uniqueNumber)
+    public void SetFee(UInt256 feeBase, UInt256 feeFactor)
+    {
+        Assert(Message.Sender == this.Owner);
+
+        var previousFeeBase = this.FeeBase;
+        var previousFeeFactor = this.FeeFactor;
+
+        this.FeeBase = feeBase;
+        this.FeeFactor = feeFactor;
+
+        Log(new FeeChangedLog { PreviousFeeBase = previousFeeBase, PreviousFeeFactor = previousFeeFactor, FeeBase = feeBase, FeeFactor = feeFactor });
+    }
+
+    public UInt256 CalculateFee(UInt256 amount)
     {
-        var template = new TransactionReferenceTemplate() { uniqueNumber = uniqueNumber, address = Message.Sender };
+        var fee = amount * this.FeeFactor / 100000000 + this.FeeBase;
+
+        // Round up to 2 decimals.
+        fee = ((fee + 999999) / 1000000) * 1000000;
+
+        return fee;
+    }
 
+    private string GetTransactionReference(UInt128 uniqueNumber, Address address)
+    {
+        var template = new TransactionReferenceTemplate() { uniqueNumber = uniqueNumber, address = address };
```

Comment:
Putting a comment here, but it is actually for `GetInvoiceReference` below.

I am not sure there is much point hashing the transaction reference to obscure it; it is possible to directly decode the parameters of the contract call onchain back into the original data as long as you know the types. Those parameters may be transactionReference itself, or the data used to derive it.

Reply:
Yes, these references are primarily intended to protect the user against thereself - i.e. so as not to submit a payment with a reference that will fail the 2nd validation.